episodes/04-muscle-couple.md

```diff
@@ -1,102 +1,338 @@
+      ports:
+        f_init: initial_state
+        o_f: final_state
+
+  conduits:
+    macro.state_out: micro.initial_state
+    micro.final_state: macro.state_in
 ```
 
-:::::::::::::::::::::::: solution 
+First, we describe the two components in this model. Components can be submodels, or helper components that convert data, control the simulation, or otherwise implement required non-model functionality. The name of a component is used by MUSCLE as an address for communication between the models. The implementation name is intended for use by a launcher, which would start the corresponding program to create an instance of a component. It is these instances that form the actual running simulation. In this example, we have two submodels: one named macro and one named micro. Macro is implemented by an implementation named diffusion, while micro is implemented by an implementation named reaction.
+
+Second, we need to connect the components together. This is done by conduits, which have a sender and a receiver. Here, we connect sending port `state_out` on component `macro` to receiving port `initial_state` on component `micro`.
+
+:::
+:::
+
+## Adding settings
 
-## Output
- 
-```output
-[1] "This new lesson looks good"
+::: instructor
+
+- Remember that we're getting our settings from MUSCLE3, they need to be added
+- It's a YAML dict, but look at the names of the settings, does anything stand out?
+
+:::
+
+The above specifies which submodels we have and how they are connected together. Next, we need to configure them by adding the settings to the yMMSL file. These will be passed to the models, who get them using the `Instance.get_settings()` function. Go ahead and add them to your `reaction_diffusion.ymmsl`:
+
+```yaml
+settings:
+  micro.t_max: 2.469136e-6
+  micro.dt: 2.469136e-8
+  macro.t_max: 1.234568e-4
+  macro.dt: 2.469136e-6
+  x_max: 1.01
+  dx: 0.01
+  k: -4.05e4    # reaction parameter
+  d: 4.05e-2    # diffusion parameter
 ```
 
-:::::::::::::::::::::::::::::::::
+Look at the names of the settings. Does anything stand out to you?
+
+## Specifying resources
 
+::: instructor
 
-## Challenge 2: how do you nest solutions within challenge blocks?
+- Resources are a bit of an advance topic, so we're not going into it here
+- We need to specify something, which is that our components are single-threaded
 
-:::::::::::::::::::::::: solution 
+:::
 
-You can add a line with at least three colons and a `solution` tag.
+Finally, we need to tell MUSCLE3 whether and if so how each model is parallelised, so that it can reserve adequate resources for each component. In this case, the models are single-threaded so that is what we specify. Again, add this to your file.
 
-:::::::::::::::::::::::::::::::::
-::::::::::::::::::::::::::::::::::::::::::::::::
+```yaml
+resources:
+  macro:
+    threads: 1
+  micro:
+    threads: 1
+```
 
-## Figures
+Note that we specify resources for each component, not for each implementation.
```

Comment:
We haven't configured the implementations yet. Should we do that as well?
```suggestion
Finally we need to tell MUSCLE where the implementations of the sub-models live for each of the two sub-models as follows:

```yaml
implementations:
  reaction:
    virtual_env: /path/to/my/env
    executable: python
    args: /path/to/my/reaction.py
```
Note that we specify resources for each component, not for each implementation.
```

Reply:
hmm, that is messed up....